grype/pkg/package.go

```diff
@@ -2,102 +2,257 @@ package pkg
+	Metadata     interface{} // This is NOT 1-for-1 the syft metadata! Only the select data needed for vulnerability matching
 }
 
 func New(p pkg.Package) Package {
+	metadataType, metadata, upstreams := dataFromPkg(p)
+
+	return Package{
+		ID:           ID(p.ID()),
+		Name:         p.Name,
+		Version:      p.Version,
+		Locations:    p.Locations,
+		Licenses:     p.Licenses,
+		Language:     p.Language,
+		Type:         p.Type,
+		CPEs:         p.CPEs,
+		PURL:         p.PURL,
+		Upstreams:    upstreams,
+		MetadataType: metadataType,
+		Metadata:     metadata,
+	}
+}
+
+func FromCatalog(catalog *pkg.Catalog) []Package {
+	result := make([]Package, 0, catalog.PackageCount())
+	for _, p := range catalog.Sorted() {
+		result = append(result, New(p))
+	}
+	return result
+}
+
+// Stringer to represent a package.
+func (p Package) String() string {
+	return fmt.Sprintf("Pkg(type=%s, name=%s, version=%s, upstreams=%d)", p.Type, p.Name, p.Version, len(p.Upstreams))
+}
+
+func dataFromPkg(p pkg.Package) (MetadataType, interface{}, []UpstreamPackage) {
 	var metadata interface{}
+	var upstreams []UpstreamPackage
+	var metadataType MetadataType
 
 	switch p.MetadataType {
 	case pkg.DpkgMetadataType:
-		if value, ok := p.Metadata.(pkg.DpkgMetadata); ok {
-			metadata = DpkgMetadata{Source: value.Source}
-		} else {
-			log.Warnf("unable to extract DPKG metadata for %s", p)
-		}
+		upstreams = dpkgDataFromPkg(p)
 	case pkg.RpmdbMetadataType:
-		if value, ok := p.Metadata.(pkg.RpmdbMetadata); ok {
-			metadata = RpmdbMetadata{SourceRpm: value.SourceRpm, Epoch: value.Epoch}
-		} else {
-			log.Warnf("unable to extract RPM metadata for %s", p)
+		m, u := rpmdbDataFromPkg(p)
+		upstreams = u
+		if m != nil {
+			metadata = *m
+			metadataType = RpmdbMetadataType
 		}
 	case pkg.JavaMetadataType:
-		if value, ok := p.Metadata.(pkg.JavaMetadata); ok {
-			var artifact, group, name string
-			if value.PomProperties != nil {
-				artifact = value.PomProperties.ArtifactID
-				group = value.PomProperties.GroupID
-			}
-			if value.Manifest != nil {
-				if n, ok := value.Manifest.Main["Name"]; ok {
-					name = n
-				}
-			}
+		if m := javaDataFromPkg(p); m != nil {
+			metadata = *m
+			metadataType = JavaMetadataType
+		}
+	case pkg.ApkMetadataType:
+		upstreams = apkDataFromPkg(p)
+	case "":
+		// let's try to extract matching-specific information from additional sources other than syft json shapes.
 
-			metadata = JavaMetadata{
-				VirtualPath:   value.VirtualPath,
-				PomArtifactID: artifact,
-				PomGroupID:    group,
-				ManifestName:  name,
+		// TODO: add java cases here
+		switch p.Type {
+		case pkg.ApkPkg:
+			upstreams = apkDataFromPURL(p.PURL)
+		case pkg.DebPkg:
+			upstreams = dpkgDataFromPURL(p.PURL)
+		case pkg.RpmPkg:
+			m, u := rpmdbDataFromPURL(p.PURL)
+			upstreams = u
+			if m != nil {
+				metadata = *m
+				metadataType = RpmdbMetadataType
 			}
-		} else {
-			log.Warnf("unable to extract Java metadata for %s", p)
 		}
```

Comment:
I'm not really sure this is the best location to deal with incoming PURLs. I've added some bits that attempt to create a Syft data model, including the metadata from the PURLs. This way I'd think Grype would be able to use a more consistent data model to populate the upstreams? See: https://github.com/anchore/syft/pull/738/files#diff-b60239c3dd2d41cfef2107e92599d7426bd918d4232e5e8273da3b05d9349bedR296

Reply:
you're right --syft is the better spot for this logic. I'll take out the specifics about deriving metadata values from pURL but keep the upstream package generation 👍